File: 2020_November_Leetcode_30_days_challenge/Week_2_Maximum Difference Between Node and Ancestor/by_dfs.py
'''

Description:

Given the root of a binary tree, find the maximum value V for which there exist different nodes A and B where V = |A.val - B.val| and A is an ancestor of B.

A node A is an ancestor of B if either: any child of A is equal to B, or any child of A is an ancestor of B.

 

Example 1:


Input: root = [8,3,10,1,6,null,14,null,null,4,7,13]
Output: 7
Explanation: We have various ancestor-node differences, some of which are given below :
|8 - 3| = 5
|3 - 7| = 4
|8 - 1| = 7
|10 - 13| = 3
Among all possible differences, the maximum value of 7 is obtained by |8 - 1| = 7.



Example 2:


Input: root = [1,null,2,null,0,3]
Output: 3
 

Constraints:

The number of nodes in the tree is in the range [2, 5000].
0 <= Node.val <= 105

Hint #1  
For each subtree, find the minimum value and maximum value of its descendants.

'''

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def maxAncestorDiff(self, root: TreeNode) -> int:
        
        def dfs( node, anc_max, anc_min ):
            
            if not node:
                
                ## base case aka stop condition
                # empty node or empty subtree
                return
            
            
            ## general case
            
            # update maximum difference between current node and ancestor
            dfs.max_diff = max(dfs.max_diff, abs(anc_max-node.val), abs(anc_min-node.val))
            
            # update maximum of ancestor
            anc_max = max(anc_max, node.val)
            
            # update minimum of ancestor
            anc_min = min(anc_min, node.val)
            
            # DFS down to next level
            dfs(node.left, anc_max, anc_min)
            dfs(node.right, anc_max, anc_min)
            
        # -------------------------------------
        
        # initialization
        dfs.max_diff = 0
        dfs(node=root, anc_max=root.val, anc_min=root.val)
        return dfs.max_diff


# n : the number of nodes in binary tree

## Time Complexity: O( n )
#
# The overhead in time is the cost of DFS, which is of O( n )

## Space Complexity: O( n )
#
# The overhead in space is the storage for recursion call stack, which is of O( n )


import unittest

class Testing( unittest.TestCase ):

    def test_case_1( self ):
        
        root = TreeNode( 8 )
        root.left = TreeNode( 3 )
        root.right = TreeNode( 10 )
        root.left.left = TreeNode( 1 )
        root.left.right = TreeNode( 6 )
        root.left.right.left = TreeNode( 4 )
        root.left.right.right = TreeNode( 7 )
        root.right.right = TreeNode( 14 )
        root.right.right.left = TreeNode( 13 )
        
        result = Solution().maxAncestorDiff( root )
        self.assertEqual(result, 7)


    def test_case_2( self ):

        root = TreeNode( 1 )
        root.right = TreeNode( 2 )
        root.right.right = TreeNode( 0 )
        root.right.right.left = TreeNode( 3 )


if __name__ == '__main__':

    unittest.main()        

File: No_0623_Add One Row to Tree/by_dfs.go
/*

Description:

Given the root of a binary tree and two integers val and depth, add a row of nodes with value val at the given depth depth.

Note that the root node is at depth 1.

The adding rule is:

Given the integer depth, for each not null tree node cur at the depth depth - 1, create two tree nodes with value val as cur's left subtree root and right subtree root.
cur's original left subtree should be the left subtree of the new left subtree root.
cur's original right subtree should be the right subtree of the new right subtree root.
If depth == 1 that means there is no depth depth - 1 at all, then create a tree node with value val as the new root of the whole original tree, and the original tree is the new root's left subtree.
 

Example 1:


Input: root = [4,2,6,3,1,5], val = 1, depth = 2
Output: [4,1,1,2,null,null,6,3,1,5]



Example 2:


Input: root = [4,2,null,3,1], val = 1, depth = 3
Output: [4,2,null,1,1,3,null,null,1]
 

Constraints:

The number of nodes in the tree is in the range [1, 104].
The depth of the tree is in the range [1, 104].
-100 <= Node.val <= 100
-105 <= val <= 105
1 <= depth <= the depth of tree + 1

*/



package main

import(

)

type TreeNode struct {
    Val int
     Left *TreeNode
    Right *TreeNode
}


func addOneRow(root *TreeNode, v int, d int) *TreeNode {
    
    if root == nil{
        
        // base case: empty tree
        
        return nil
        
    } else if d == 1{
        
        // base case: add one row above original root
        
        return &TreeNode{v, root, nil}
        
    } else if d == 2{
        
        // base case: add one row just below original root
        
        root.Left = &TreeNode{v, root.Left, nil}
        root.Right = &TreeNode{v, nil, root.Right}
        
        return root
        
    } else{
        
        // general case: depth >= 3
        // do it in DFS with common pattern
        
        root.Left = addOneRow(root.Left, v, d-1)
        root.Right = addOneRow(root.Right, v, d-1)
        
        return root
    }
    
    
    
}
//end of function addOneRow



// n : the number of nodes

//// Time Complexity: O( n )
//
// The overhead in time is the cost of DFS, which is of O( n )

//// Space Complexity: O( n )
//
// The overhead in space is the storage for recursion, which is of O( n )

// type "go test -v" in console to run unittest

